### ISSUE TITLE
26.0.90; Undoing yasnippet expansion bug

### ISSUE THREAD

**Issue by deadtrickster** [joaotavora/yasnippet#869]
This what I sent to emacs-bugs. It didn't get through so I first reported it in #821 and was asked to create
a separate issue.

![yasnippet-undo-bug](https://user-images.githubusercontent.com/172311/32936662-2e506d2e-cb86-11e7-9bc3-85e1ff0a7d18.gif)

In GNU Emacs 26.0.90
Repository revision: 1faade882165de2e9bd63702c59c8f26531c18ed

Undoing yasnippet expansion results in weird state.
Also saw this on previous Emacs versions.

I attached gif that illustrates the bug. Major mode is Erlang.

```
erlang is an installed package.

     Status: Installed in ‘erlang-20170913.604/’ (unsigned). Delete
    Version: 20170913.604
    Summary: Erlang major mode
   Requires: emacs-24.1
Other versions: 20170913.604 (melpa).
```

Snippet is todo: https://github.com/AndreaCrotti/yasnippet-snippets/blob/3226ab8d879d2b892ec0a269d0cca4ae086aa7c3/prog-mode/todo.

What's interesting:

Todo tries to expand even inside words, e.g. (illustrating cursor with |):

```
-spec open(inet:hostname(), config()) -> t|erm().
open(Host, Config) ->
  ...
  ok.

-spec open(inet:hostname(), config()) -> %TODO: erm().
open(Host, Config) ->
  ...
  ok.
```

but

```
port|:qwe() => port:qwe() 
```

now this (not valid erlang, but looks like multi line expression):

```
t|erm()
end.

=> (expand, undo)

tend.
```

One more example:

```
t|erm().
qwe ->
end.
    
=> (expand, undo)

t
qwe ->
end.
```

The same behaviour observed for other "commenting" snippets from prog-mode collection.

Thank you,
Ilya

**Comment by npostavs** [comment]
There seems to be at least two bugs here:

1. The first line of the snippet is indented, even though `yas-also-indent-first-line` is `nil`.
2. We pick up some undo info from the indentation, leading to incorrect result whening undoing snippet expansion.

**Comment by npostavs** [comment]
Actually the bug is not erlang-mode specific, the following test fails:
```elisp
(ert-deftest undo-indentation-2 ()
  "Check undoing works when only line of snippet is indented."
  (let ((yas-also-auto-indent-first-line t)
        (indent-tabs-mode nil))
    (yas-with-snippet-dirs
     '((".emacs.d/snippets" ("emacs-lisp-mode" ("t" . "; TODO"))))
     (with-temp-buffer
       (emacs-lisp-mode)
       (yas-reload-all)
       (yas-minor-mode 1)
       (insert "t")
       (setq buffer-undo-list ())
       (ert-simulate-command '(yas-expand))
       ;; Need undo barrier, I think command loop puts it normally.
       (push nil buffer-undo-list)
       (should (string= (buffer-string) (concat (make-string comment-column ?\s) "; TODO")))
       (ert-simulate-command '(undo))
       (should (string= (buffer-string) "t"))))))
```

It's fixed by removing this code in `yas-expand-snippet`, but then the existing `undo-indentation` fails...
```elisp
           ;; Undo actions from indent of snippet's 1st line.
           (setq buffer-undo-list
                 (nconc yas--first-indent-undo buffer-undo-list))
```

**Comment by npostavs** [comment]
Okay, I think I've got it this time, see #871.

**Comment by deadtrickster** [comment]
tried #871, works! Thanks.

Still can't find how to disable t|erm() -> %TODO: erm(), i.e. inside words. I think `yas--templates-for-key-at-point` just unconditionally does that. But that's not related to the bug(s) fixed there, right?

**Comment by npostavs** [comment]
> Still can't find how to disable t|erm() -> %TODO: erm(), i.e. inside words. I think yas--templates-for-key-at-point just unconditionally does that. But that's not related to the bug(s) fixed there, right?

Right. That is somewhat surprising behaviour, but somehow I don't think anyone ever mentioned it before. `yas-key-syntaxes` controls expansion based on what's *before* point, but nothing checks what is after it. You could use `yas-buffer-local-condition` to change this: 
```elisp
(setq yas-buffer-local-condition '(looking-at "\\>"))
```

**Comment by deadtrickster** [comment]
Awesome, thanks again!
